src/Tizen.Applications.ComponentBased/Tizen.Applications.ComponentBased.Common/ComponentBasedApplication.cs

```diff
@@ -0,0 +1,209 @@
+        /// <since_tizen> 6 </since_tizen>
+        public void RegisterComponent(Type compType, string compId)
+        {
+            if (_componentFactories.ContainsKey(compType))
+            {
+                throw new InvalidOperationException("Already exist type");
+            }
+
+            if (typeof(FrameComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add frame component");
+                _componentFactories.Add(compType, new FrameFactory(compType, compId, this));
+            }
+            else if (typeof(ServiceComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add service component");
+                _componentFactories.Add(compType, new ServiceFactory(compType, compId, this));
+            }
+            else if (typeof(BaseComponent).IsAssignableFrom(compType))
+            {
+                Log.Info(LogTag, "Add base component");
+                BaseComponent b = Activator.CreateInstance(compType) as BaseComponent;
+                _componentFactories.Add(compType, new BaseFactory(compType, compId, b.ComponentType, this));
+            }
+            else
+            {
+                throw new ArgumentException("Invalid type");
+            }
+        }
+
+        /// <summary>
+        /// Runs the application's main loop.
+        /// </summary>
+        /// <param name="args">Arguments from commandline.</param>
+        /// <exception cref="InvalidOperationException">Thrown when component type is already added to the component.</exception>
+        /// <since_tizen> 6 </since_tizen>
+        public override void Run(string[] args)
+        {
+            base.Run(args);
+
+            string[] argsClone = new string[args.Length + 1];
+            if (args.Length > 1)
+            {
+                args.CopyTo(argsClone, 1);
+            }
+            argsClone[0] = string.Empty;
+
+            Interop.CBApplication.ErrorCode err = Interop.CBApplication.BaseMain(argsClone.Length, argsClone, ref _callbacks, IntPtr.Zero);
+            if (err != Interop.CBApplication.ErrorCode.None)
+            {
+                Log.Error(LogTag, "Failed to run the application. Err = " + err);
+                throw new InvalidOperationException("Fail to run application : err(" + err + ")");
+            }
+        }
+
+        /// <summary>
+        /// Exits the main loop of the application.
+        /// </summary>
+        /// <since_tizen> 6 </since_tizen>
+        public override void Exit()
+        {
+            Interop.CBApplication.BaseExit();
+        }
+
+        private IntPtr OnCreateNative(IntPtr data)
+        {
+            Log.Debug(LogTag, "On create");
+            IntPtr h = IntPtr.Zero;
+            foreach(KeyValuePair<Type, ComponentFactoryBase> entry in _componentFactories)
+            {
+                h = entry.Value.Bind(h);
+            }
+
+            return h;
```

Comment:
What means of return value?
It is only returning a last returned value from Bind
Is that ok?

Reply:
Yes, returned value contains native-handle that contains native-component-factory map.
Bind function adds native-component-factory to the handle, therefore returned value contains multiple native-component-factory.

For further information, you can see native code here. https://review.tizen.org/gerrit/#/admin/projects/platform/core/appfw/component-based-application

Bind function will call native API component_based_app_base_add_frame_component.